File: star-proto/tests/typescript/src/serialization.test.ts
// serialization.test.ts - TypeScript Protobuf Serialization Tests
// Verifies that generated TypeScript code serializes/deserializes correctly.
//
// STAR Project - Texas A&M University
// December 2025

import { describe, it, expect } from 'vitest';

// Import generated classes (will be available after buf generate)
// import { VelocityCommand } from '@gen/star/v1/motor_control';
// import { TelemetryData, ImuData } from '@gen/star/v1/telemetry';
// import { Status, ConnectionStatus, RobotMode } from '@gen/star/v1/common';

describe('Protobuf Serialization Tests', () => {

    describe('Test Framework', () => {
        it('should have working test framework', () => {
            // Placeholder test - verifies Vitest is operational
            expect(true).toBe(true);
        });
    });

    // TODO: Uncomment after buf generate produces TypeScript code
    /*
    describe('VelocityCommand - 4 Motors', () => {
        it('should round-trip serialize correctly with 4 motors', () => {
            // Create command with 4 independent motors (MKS units)
            const original: VelocityCommand = {
                motor0VelocityMps: 1.5,
                motor1VelocityMps: 1.5,
                motor2VelocityMps: 1.0,
                motor3VelocityMps: 1.0,
                sequence: 42,
                timestampUs: BigInt(Date.now() * 1000),
            };

            // Serialize to binary
            const bytes = VelocityCommand.toBinary(original);

            // Verify compact encoding
            expect(bytes.length).toBeLessThan(100);

            // Deserialize
            const deserialized = VelocityCommand.fromBinary(bytes);

            // Verify all 4 motor fields
            expect(deserialized.motor0VelocityMps).toBeCloseTo(original.motor0VelocityMps, 3);
            expect(deserialized.motor1VelocityMps).toBeCloseTo(original.motor1VelocityMps, 3);
            expect(deserialized.motor2VelocityMps).toBeCloseTo(original.motor2VelocityMps, 3);
            expect(deserialized.motor3VelocityMps).toBeCloseTo(original.motor3VelocityMps, 3);
            expect(deserialized.sequence).toBe(original.sequence);
        });

        it('should support differential drive mode', () => {
            const leftVel = 2.0;
            const rightVel = 1.5;

            const command: VelocityCommand = {
                motor0VelocityMps: leftVel,  // Left front
                motor1VelocityMps: leftVel,  // Left rear
                motor2VelocityMps: rightVel, // Right front
                motor3VelocityMps: rightVel, // Right rear
                sequence: 1,
                timestampUs: BigInt(0),
            };

            const json = VelocityCommand.toJsonString(command);
            const parsed = VelocityCommand.fromJsonString(json);

            // Verify left motors match
            expect(parsed.motor0VelocityMps).toBeCloseTo(leftVel, 3);
            expect(parsed.motor1VelocityMps).toBeCloseTo(leftVel, 3);

            // Verify right motors match
            expect(parsed.motor2VelocityMps).toBeCloseTo(rightVel, 3);
            expect(parsed.motor3VelocityMps).toBeCloseTo(rightVel, 3);
        });
    });

    describe('EncoderData Streaming - 4 Motors', () => {
        it('should handle streaming message serialization for all 4 motors', () => {
            const messages: EncoderData[] = [];

            // Create multiple encoder data messages (test all 4 motors)
            for (let i = 0; i < 12; i++) {
                messages.push({
                    motorId: i % 4,  // Motor IDs 0, 1, 2, 3
                    ticks: BigInt(i * 1000),
                    velocityMps: i * 0.1,
                    timestampUs: BigInt(Date.now() * 1000),
                });
            }

            // Serialize each message
            const serialized = messages.map(m => EncoderData.toBinary(m));

            // Deserialize and verify
            serialized.forEach((bytes, i) => {
                const data = EncoderData.fromBinary(bytes);
                expect(data.motorId).toBe(i % 4);
                expect(Number(data.ticks)).toBe(i * 1000);
            });
        });
    });

    describe('Enum Values', () => {
        it('should have UNKNOWN as zero value (Boston Dynamics style)', () => {
            expect(Status.STATUS_UNKNOWN).toBe(0);
            expect(ConnectionStatus.CONNECTION_STATUS_UNKNOWN).toBe(0);
            expect(RobotMode.ROBOT_MODE_UNKNOWN).toBe(0);
        });

        it('should preserve enum values through serialization', () => {
            const response = {
                requestId: 'enum-test',
                status: Status.STATUS_OK,
                errorMessage: '',
                latencyUs: BigInt(0),
            };

            const bytes = ResponseHeader.toBinary(response);
            const parsed = ResponseHeader.fromBinary(bytes);

            expect(parsed.status).toBe(Status.STATUS_OK);
        });
    });

    describe('Unit Conventions', () => {
        it('should use MKS units for all measurements', () => {
            const telemetry: TelemetryData = {
                imu: {
                    pitchRad: Math.PI / 4,       // radians
                    rollRad: 0,
                    yawRad: Math.PI,
                    accelXMps2: 0,
                    accelYMps2: 0,
                    accelZMps2: 9.81,           // m/s^2
                    gyroXRadPerS: 0.1,          // rad/s
                    gyroYRadPerS: 0,
                    gyroZRadPerS: 0,
                },
                batteryPercent: 75.0,
                temperatureCelsius: 35.0,
            };

            const bytes = TelemetryData.toBinary(telemetry);
            const parsed = TelemetryData.fromBinary(bytes);

            // Verify MKS units preserved
            expect(parsed.imu?.accelZMps2).toBeCloseTo(9.81, 2);
            expect(parsed.imu?.pitchRad).toBeCloseTo(Math.PI / 4, 4);
        });
    });

    describe('RequestHeader/ResponseHeader Pattern', () => {
        it('should support header pattern for all requests', () => {
            const requestId = `test-${Date.now()}`;

            const request: SetVelocityRequest = {
                header: {
                    requestId: requestId,
                    clientVersion: '1.0.0',
                },
                command: {
                    // Differential drive mode: all motors at 1.0 m/s (straight forward)
                    motor0VelocityMps: 1.0,  // Left front
                    motor1VelocityMps: 1.0,  // Left rear
                    motor2VelocityMps: 1.0,  // Right front
                    motor3VelocityMps: 1.0,  // Right rear
                    sequence: 1,
                    timestampUs: BigInt(Date.now() * 1000),
                },
            };

            const bytes = SetVelocityRequest.toBinary(request);
            const parsed = SetVelocityRequest.fromBinary(bytes);

            expect(parsed.header?.requestId).toBe(requestId);
        });
    });
    */
});

File: star-proto/tests/go/serialization_test.go
// serialization_test.go - STAR Protobuf Serialization Tests
// Verifies protobuf round-trip serialization for all message types.
//
// Using Go table-driven tests following standard patterns.
//
// STAR Project - Texas A&M University
// December 2025

package starproto_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/proto"

	starv1 "github.com/Locked-Inc/star-proto/gen/go/star/v1"
)

// ============================================================================
// Common Message Tests
// ============================================================================

func TestRequestHeaderRoundTrip(t *testing.T) {
	tests := []struct {
		name          string
		requestID     string
		clientVersion string
	}{
		{"basic", "req-001", "1.0.0"},
		{"uuid", "550e8400-e29b-41d4-a716-446655440000", "2.1.3"},
		{"empty_version", "req-002", ""},
		{"long_id", "very-long-request-identifier-for-testing-purposes", "1.0.0-beta"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			header := &starv1.RequestHeader{
				RequestId:     tc.requestID,
				ClientVersion: tc.clientVersion,
			}

			bytes, err := proto.Marshal(header)
			require.NoError(t, err, "Marshal should succeed")

			parsed := &starv1.RequestHeader{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err, "Unmarshal should succeed")

			assert.Equal(t, tc.requestID, parsed.RequestId)
			assert.Equal(t, tc.clientVersion, parsed.ClientVersion)
		})
	}
}

func TestResponseHeaderRoundTrip(t *testing.T) {
	tests := []struct {
		name      string
		requestID string
		status    starv1.Status
	}{
		{"ok", "req-001", starv1.Status_STATUS_OK},
		{"error", "req-002", starv1.Status_STATUS_INTERNAL_ERROR},
		{"invalid", "req-003", starv1.Status_STATUS_INVALID_REQUEST},
		{"timeout", "req-004", starv1.Status_STATUS_TIMEOUT},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			header := &starv1.ResponseHeader{
				RequestId: tc.requestID,
				Status:    tc.status,
			}

			bytes, err := proto.Marshal(header)
			require.NoError(t, err)

			parsed := &starv1.ResponseHeader{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err)

			assert.Equal(t, tc.requestID, parsed.RequestId)
			assert.Equal(t, tc.status, parsed.Status)
		})
	}
}

func TestStatusEnumZeroValue(t *testing.T) {
	// Boston Dynamics style: zero value must be UNKNOWN
	assert.Equal(t, int32(0), int32(starv1.Status_STATUS_UNKNOWN))
}

func TestAllStatusEnumValues(t *testing.T) {
	statuses := []starv1.Status{
		starv1.Status_STATUS_UNKNOWN,
		starv1.Status_STATUS_OK,
		starv1.Status_STATUS_INVALID_REQUEST,
		starv1.Status_STATUS_INTERNAL_ERROR,
		starv1.Status_STATUS_NOT_FOUND,
		starv1.Status_STATUS_TIMEOUT,
		starv1.Status_STATUS_INVALID_STATE,
		starv1.Status_STATUS_ESTOP_ACTIVE,
	}

	for _, status := range statuses {
		t.Run(status.String(), func(t *testing.T) {
			header := &starv1.ResponseHeader{
				RequestId: "test",
				Status:    status,
			}

			bytes, err := proto.Marshal(header)
			require.NoError(t, err)

			parsed := &starv1.ResponseHeader{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err)

			assert.Equal(t, status, parsed.Status)
		})
	}
}

// ============================================================================
// Motor Control Tests
// ============================================================================

func TestVelocityCommandRoundTrip(t *testing.T) {
	tests := []struct {
		name     string
		leftMps  float64
		rightMps float64
		sequence uint32
		desc     string
	}{
		{"forward_full", 2.0, 2.0, 1, "Both wheels forward at max speed"},
		{"reverse_full", -2.0, -2.0, 2, "Both wheels reverse at max speed"},
		{"stop", 0.0, 0.0, 3, "Both wheels stopped"},
		{"turn_left", 0.5, 1.5, 4, "Left turn - right wheel faster"},
		{"turn_right", 1.5, 0.5, 5, "Right turn - left wheel faster"},
		{"spin_left", -1.0, 1.0, 6, "Spin in place counterclockwise"},
		{"spin_right", 1.0, -1.0, 7, "Spin in place clockwise"},
		{"precision_low", 0.001, 0.001, 8, "Very low velocity for precision"},
		{"asymmetric", 1.234, -0.567, 9, "Asymmetric decimal velocities"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			cmd := &starv1.VelocityCommand{
				LeftVelocityMps:  tc.leftMps,
				RightVelocityMps: tc.rightMps,
				Sequence:         tc.sequence,
			}

			bytes, err := proto.Marshal(cmd)
			require.NoError(t, err, tc.desc)

			parsed := &starv1.VelocityCommand{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err, tc.desc)

			assert.InDelta(t, tc.leftMps, parsed.LeftVelocityMps, 0.0001, tc.desc)
			assert.InDelta(t, tc.rightMps, parsed.RightVelocityMps, 0.0001, tc.desc)
			assert.Equal(t, tc.sequence, parsed.Sequence, tc.desc)
		})
	}
}

func TestEncoderDataRoundTrip(t *testing.T) {
	tests := []struct {
		name        string
		motorID     int32
		ticks       int64
		velocityMps float64
	}{
		{"motor0_rest", 0, 0, 0.0},
		{"motor0_forward", 0, 1000, 0.5},
		{"motor0_reverse", 0, -1000, -0.5},
		{"motor1_forward", 1, 5000, 1.2},
		{"motor2_forward", 2, 10000, 2.0},
		{"motor3_forward", 3, -5000, -1.2},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			encoder := &starv1.EncoderData{
				MotorId:     tc.motorID,
				Ticks:       tc.ticks,
				VelocityMps: tc.velocityMps,
			}

			bytes, err := proto.Marshal(encoder)
			require.NoError(t, err)

			parsed := &starv1.EncoderData{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err)

			assert.Equal(t, tc.motorID, parsed.MotorId)
			assert.Equal(t, tc.ticks, parsed.Ticks)
			assert.InDelta(t, tc.velocityMps, parsed.VelocityMps, 0.0001)
		})
	}
}

func TestPidConfigRoundTrip(t *testing.T) {
	tests := []struct {
		name string
		kp   float64
		ki   float64
		kd   float64
		desc string
	}{
		{"matlab_tuned", 0.286, 8.01, 0.0, "MATLAB-tuned gains for velocity control"},
		{"conservative", 0.1, 1.0, 0.01, "Conservative gains for stability"},
		{"aggressive", 2.0, 20.0, 0.5, "Aggressive gains for fast response"},
		{"p_only", 1.0, 0.0, 0.0, "Proportional-only control"},
		{"pi_control", 0.5, 5.0, 0.0, "PI control without derivative"},
		{"pd_control", 1.0, 0.0, 0.2, "PD control without integral"},
		{"disabled", 0.0, 0.0, 0.0, "All gains zero - motor disabled"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			pid := &starv1.PidConfig{
				Kp: tc.kp,
				Ki: tc.ki,
				Kd: tc.kd,
			}

			bytes, err := proto.Marshal(pid)
			require.NoError(t, err, tc.desc)

			parsed := &starv1.PidConfig{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err, tc.desc)

			assert.InDelta(t, tc.kp, parsed.Kp, 0.0001, tc.desc)
			assert.InDelta(t, tc.ki, parsed.Ki, 0.0001, tc.desc)
			assert.InDelta(t, tc.kd, parsed.Kd, 0.0001, tc.desc)
		})
	}
}

// ============================================================================
// Telemetry Tests
// ============================================================================

func TestImuDataRoundTrip(t *testing.T) {
	tests := []struct {
		name    string
		accelX  float64
		accelY  float64
		accelZ  float64
		gyroX   float64
		gyroY   float64
		gyroZ   float64
		desc    string
	}{
		{"rest", 0.0, 0.0, 9.81, 0.0, 0.0, 0.0, "At rest, gravity on Z"},
		{"tilted", 4.9, 0.0, 8.5, 0.0, 0.0, 0.0, "Tilted 30 degrees"},
		{"rotating", 0.0, 0.0, 9.81, 0.0, 0.0, 1.57, "Rotating around Z"},
		{"moving", 2.0, 1.0, 9.81, 0.1, 0.05, 0.2, "Moving with rotation"},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			imu := &starv1.ImuData{
				AccelXMps2:   tc.accelX,
				AccelYMps2:   tc.accelY,
				AccelZMps2:   tc.accelZ,
				GyroXRadPerS: tc.gyroX,
				GyroYRadPerS: tc.gyroY,
				GyroZRadPerS: tc.gyroZ,
			}

			bytes, err := proto.Marshal(imu)
			require.NoError(t, err, tc.desc)

			parsed := &starv1.ImuData{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err, tc.desc)

			assert.InDelta(t, tc.accelX, parsed.AccelXMps2, 0.0001, tc.desc)
			assert.InDelta(t, tc.accelY, parsed.AccelYMps2, 0.0001, tc.desc)
			assert.InDelta(t, tc.accelZ, parsed.AccelZMps2, 0.0001, tc.desc)
			assert.InDelta(t, tc.gyroX, parsed.GyroXRadPerS, 0.0001, tc.desc)
			assert.InDelta(t, tc.gyroY, parsed.GyroYRadPerS, 0.0001, tc.desc)
			assert.InDelta(t, tc.gyroZ, parsed.GyroZRadPerS, 0.0001, tc.desc)
		})
	}
}

func TestSystemStatusRoundTrip(t *testing.T) {
	tests := []struct {
		name            string
		connStatus      starv1.ConnectionStatus
		mode            starv1.RobotMode
		esp32Connected  bool
		lidarConnected  bool
		rosConnected    bool
		firmwareVersion string
		uptimeS         uint64
	}{
		{"startup", starv1.ConnectionStatus_CONNECTION_STATUS_UNKNOWN, starv1.RobotMode_ROBOT_MODE_IDLE, false, false, false, "", 0},
		{"connected", starv1.ConnectionStatus_CONNECTION_STATUS_CONNECTED, starv1.RobotMode_ROBOT_MODE_MANUAL, true, true, true, "1.0.0", 3600},
		{"autonomous", starv1.ConnectionStatus_CONNECTION_STATUS_CONNECTED, starv1.RobotMode_ROBOT_MODE_AUTONOMOUS, true, true, true, "2.1.0", 86400},
		{"partial", starv1.ConnectionStatus_CONNECTION_STATUS_CONNECTING, starv1.RobotMode_ROBOT_MODE_IDLE, true, false, false, "1.0.0-beta", 100},
		{"disconnected", starv1.ConnectionStatus_CONNECTION_STATUS_DISCONNECTED, starv1.RobotMode_ROBOT_MODE_EMERGENCY_STOP, false, false, false, "1.0.0", 7200},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			status := &starv1.SystemStatus{
				ConnectionStatus: tc.connStatus,
				Mode:             tc.mode,
				Esp32Connected:   tc.esp32Connected,
				LidarConnected:   tc.lidarConnected,
				RosConnected:     tc.rosConnected,
				FirmwareVersion:  tc.firmwareVersion,
				UptimeS:          tc.uptimeS,
			}

			bytes, err := proto.Marshal(status)
			require.NoError(t, err)

			parsed := &starv1.SystemStatus{}
			err = proto.Unmarshal(bytes, parsed)
			require.NoError(t, err)

			assert.Equal(t, tc.connStatus, parsed.ConnectionStatus)
			assert.Equal(t, tc.mode, parsed.Mode)
			assert.Equal(t, tc.esp32Connected, parsed.Esp32Connected)
			assert.Equal(t, tc.lidarConnected, parsed.LidarConnected)
			assert.Equal(t, tc.rosConnected, parsed.RosConnected)
			assert.Equal(t, tc.firmwareVersion, parsed.FirmwareVersion)
			assert.Equal(t, tc.uptimeS, parsed.UptimeS)
		})
	}
}

// ============================================================================
// Request/Response Pattern Tests
// ============================================================================

func TestSetVelocityRequestResponse(t *testing.T) {
	// Request
	req := &starv1.SetVelocityRequest{
		Header: &starv1.RequestHeader{
			RequestId:     "vel-001",
			ClientVersion: "1.0.0",
		},
		Command: &starv1.VelocityCommand{
			LeftVelocityMps:  1.5,
			RightVelocityMps: 1.5,
			Sequence:         42,
		},
	}

	reqBytes, err := proto.Marshal(req)
	require.NoError(t, err)

	parsedReq := &starv1.SetVelocityRequest{}
	err = proto.Unmarshal(reqBytes, parsedReq)
	require.NoError(t, err)

	assert.Equal(t, "vel-001", parsedReq.Header.RequestId)
	assert.InDelta(t, 1.5, parsedReq.Command.LeftVelocityMps, 0.0001)

	// Response
	resp := &starv1.SetVelocityResponse{
		Header: &starv1.ResponseHeader{
			RequestId: "vel-001",
			Status:    starv1.Status_STATUS_OK,
		},
	}

	respBytes, err := proto.Marshal(resp)
	require.NoError(t, err)

	parsedResp := &starv1.SetVelocityResponse{}
	err = proto.Unmarshal(respBytes, parsedResp)
	require.NoError(t, err)

	assert.Equal(t, "vel-001", parsedResp.Header.RequestId)
	assert.Equal(t, starv1.Status_STATUS_OK, parsedResp.Header.Status)
}

// ============================================================================
// Streaming Simulation Tests
// ============================================================================

func TestEncoderStreamSimulation(t *testing.T) {
	tests := []struct {
		name  string
		count int
	}{
		{"1_message", 1},
		{"10_messages", 10},
		{"100_messages", 100},
		{"1000_messages", 1000},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			var totalSize int
			for i := 0; i < tc.count; i++ {
				encoder := &starv1.EncoderData{
					MotorId:     int32(i % 4),
					Ticks:       int64(i * 100),
					VelocityMps: float64(i) * 0.01,
				}

				bytes, err := proto.Marshal(encoder)
				require.NoError(t, err)
				totalSize += len(bytes)

				parsed := &starv1.EncoderData{}
				err = proto.Unmarshal(bytes, parsed)
				require.NoError(t, err)
			}

			t.Logf("Streamed %d messages, total size: %d bytes", tc.count, totalSize)
		})
	}
}
